### PR TITLE
feat: add te_fused_topk_with_score_function backward

### PR DESCRIPTION
**Pull request by douxetpur**
### PR Category

Operator, OP Test, Benchmark

### Type of Change

New Feature

### Description

This PR adds a Triton implementation of `te_fused_topk_with_score_function_bwd`, the backward operator for fused MoE top-k routing.

The implementation follows NVIDIA TransformerEngine semantics and supports:

- Sigmoid, softmax, and sqrtsoftplus score functions.
- Pre-top-k and post-top-k softmax modes.
- Backward propagation of the routing scaling factor.
- Normalization backward for sigmoid and sqrtsoftplus when `topk > 1`.
- FP32 intermediate computation with output converted back to the input gradient dtype.
- FP16, BF16, and FP32 input gradients.

The new operator is also exported through `flag_gems.fused`.

PyTorch-reference-based unit tests cover:

- All supported score functions.
- Pre-top-k and post-top-k softmax.
- Multiple token, expert, and top-k configurations.
- FP16, BF16, and FP32.
- Different routing scaling factors.

Additional tests compare the Triton implementation against NVIDIA TransformerEngine native CUDA implementation for:

- Sigmoid.
- Pre-top-k softmax.
- Post-top-k softmax.
- FP16, BF16, and FP32.

`sqrtsoftplus` vs TransformerEngine is currently deferred because the matching forward operator is not available yet. It remains covered by the PyTorch reference tests in this PR.

### Performance

A benchmark has been added for `te_fused_topk_with_score_function_bwd`.

It compares the FlagGems Triton implementation with NVIDIA TransformerEngine native CUDA implementation for:

- Sigmoid.
- Pre-top-k softmax.
- Post-top-k softmax.
- FP16, BF16, and FP32.
- Token counts from 1 to 4096.
- Expert counts from 64 to 256.
- `topk = 8`.

When TransformerEngine is not installed, the benchmark falls back to the PyTorch reference implementation and also includes `sqrtsoftplus`.

Forward intermediates and routing maps are prepared before timing, so the reported latency measures only the backward operation.

Note: the benchmark framework reports the baseline column as `Torch Latency`. In this benchmark, that column means TransformerEngine native latency when TransformerEngine is available, otherwise PyTorch reference latency.

Benchmark command:

```bash
python -m pytest -v -s -rs benchmark/test_te_fused_topk_with_score_function_bwd.py --level comprehensive
```

Benchmark result summary:

| Score function | DType | TE Native Latency Range (ms) | FlagGems Latency Range (ms) | Speedup Range |
| --- | --- | --- | --- | --- |
| Sigmoid | FP16 | 0.012320 - 0.065856 | 0.005280 - 0.010560 | 2.33x - 6.24x |
| Sigmoid | BF16 | 0.012256 - 0.067968 | 0.005120 - 0.010496 | 2.39x - 6.48x |
| Sigmoid | FP32 | 0.008000 - 0.055456 | 0.005024 - 0.011392 | 1.52x - 4.87x |
| Softmax pre-top-k | FP16 | 0.010272 - 0.032048 | 0.004864 - 0.009440 | 2.11x - 3.40x |
| Softmax pre-top-k | BF16 | 0.010336 - 0.031936 | 0.004992 - 0.009632 | 2.07x - 3.32x |
| Softmax pre-top-k | FP32 | 0.006336 - 0.019840 | 0.005024 - 0.010656 | 1.19x - 1.86x |
| Softmax post-top-k | FP16 | 0.010560 - 0.031328 | 0.004832 - 0.009536 | 2.19x - 3.29x |
| Softmax post-top-k | BF16 | 0.010592 - 0.031520 | 0.004832 - 0.009760 | 2.19x - 3.23x |
| Softmax post-top-k | FP32 | 0.006560 - 0.021152 | 0.005024 - 0.010560 | 1.26x - 2.00x |

`sqrtsoftplus` benchmark against TransformerEngine is skipped for now and will be enabled after the matching forward operator is added.

### CODE REVIEW

File: benchmark/test_fused_topk_with_score_function_bwd.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
+# Copyright 2026 FlagOS Contributors
```
Comment (0x45f) [inline benchmark/test_fused_topk_with_score_function_bwd.py]: 1. Add test cases comparing against Te in correctness and performance testing. 
2. Add op info in conf/operators.yaml

Comment (douxetpur) [reply benchmark/test_fused_topk_with_score_function_bwd.py]: Thank for your suggestion.

Comment (douxetpur) [reply benchmark/test_fused_topk_with_score_function_bwd.py]: I added correctness tests and benchmark cases comparing the FlagGems implementation against TransformerEngine native CUDA implementation when TE is available. If TE is not installed, the tests/benchmark fall back to the PyTorch reference implementation.

I also added the operator info in `conf/operators.yaml`.

Note: `sqrtsoftplus` vs TE is deferred until the matching forward operator is added, and is currently covered by the PyTorch reference tests.

File: tests/test_te_fused_topk_with_score_function_bwd.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+    grad_probs = torch.randn(num_tokens, num_experts, device=device, dtype=dtype)
+
+    return routing_map, intermediate, grad_probs
+
+
+@pytest.mark.parametrize("num_tokens", [1, 16, 128])
+@pytest.mark.parametrize("num_experts", [8, 64, 256])
+@pytest.mark.parametrize("topk", [1, 2, 8])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16, torch.float32])
+class TestFusedTopkScoreFnBwdSigmoid:
```
Comment (0x45f) [inline tests/test_te_fused_topk_with_score_function_bwd.py:146]: why using class

File: tests/test_te_fused_topk_with_score_function_bwd.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+
+    intermediate = intermediate.float()
+
+    # grad_probs
+    grad_probs = torch.randn(num_tokens, num_experts, device=device, dtype=dtype)
+
+    return routing_map, intermediate, grad_probs
+
+
+@pytest.mark.parametrize("num_tokens", [1, 16, 128])
```
Comment (0x45f) [inline tests/test_te_fused_topk_with_score_function_bwd.py:142]: add mark.op_name

File: tests/test_te_fused_topk_with_score_function_bwd.py [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+            topk=topk,
+            scaling_factor=scaling_factor,
+            score_function=0,
+        )
+
+        expected = _reference_sigmoid_bwd(
+            routing_map, intermediate, grad_probs, topk, scaling_factor
+        )
+
+        torch.testing.assert_close(result, expected, rtol=1e-2, atol=1e-2)
```
Comment (0x45f) [inline tests/test_te_fused_topk_with_score_function_bwd.py:171]: using gems_assert_close
